imap/http_proxy.c

```diff
@@ -1243,3 +1261,212 @@ EXPORTED int http_proxy_copy(struct backend *src_be, struct backend *dest_be,
+                txn->req_uri->path, HTTP_VERSION,
+                be->hostname, CYRUS_VERSION);
+    write_forwarding_hdrs(&be_txn, txn->req_hdrs, txn->req_line.ver,
+                          https ? "https" : "http");
+    prot_puts(be->out, "Connection: Upgrade\r\n");
+    hdr = spool_getheader(txn->req_hdrs, ":protocol");
+    if (hdr && *hdr) {
+        prot_printf(be->out, "Upgrade: %s\r\n", *hdr);
+    }
+    prot_printf(be->out, "Sec-WebSocket-Key: %s\r\n",
+                "Q3lydXMgSFRUUCBQcm94eQ==");  // "Cyrus HTTP Proxy" b64-encoded
+    spool_enum_hdrcache(txn->req_hdrs, &write_cachehdr, &be_txn);
+    prot_puts(be->out, "\r\n");
+    prot_flush(be->out);
+    buf_free(&be_txn.buf);
+
+    /* Read response from backend */
+    resp_body.flags = 0;
+
+    r = http_read_response(be, METH_GET, &code,
+                           &resp_hdrs, &resp_body, &txn->error.desc);
+    if (r || (resp_body.flags & BODY_CLOSE)) {
+        proxy_downserver(be);
+        goto done;
+    }
+
+    /* Send response to client */
+    if (code == 101) {
+        code = 200;
+        txn->flags.te = TE_CHUNKED;
+    }
+    http_err = http_status_to_code(code);
+    send_response(txn, http_err, resp_hdrs, &resp_body.payload);
+
+    log_proxy_request(http_err, txn, resp_hdrs, &resp_body);
+
+ done:
+    buf_free(&resp_body.payload);
+    if (resp_hdrs) spool_free_hdrcache(resp_hdrs);
+
+    return r;
+}
+
+/*
+ * Check an array of streams for input.
+ *
+ * Input from serverin is sent to clientout.
+ * If serverout is non-NULL:
+ *   - input from clientin is sent to serverout.
+ *   - returns -1 if clientin or serverin closed, otherwise returns 0.
+ * If serverout is NULL:
+ *   - returns 1 if input from clientin is pending, otherwise returns 0.
+ */
+EXPORTED int http_proxy_check_input(struct http_connection *conn,
+                                    ptrarray_t *pipes,
+                                    unsigned long timeout_sec)
+{
+    struct protgroup *protin = conn->pgin;
+    struct protgroup *protout = NULL;
+    struct timeval timeout = { timeout_sec, 0 };
+    struct protstream *clientin = conn->pin;
+    struct protstream *clientout = conn->pout;
+    struct protstream *serverout = NULL;
+    struct transaction_t *txn = NULL;
+    int n, ret = 0;
+
+    protgroup_reset(protin);
+    protgroup_insert(protin, clientin);
+
+    for (n = 0; n < ptrarray_size(pipes); n++) {
+        txn = ptrarray_nth(pipes, n);
+        protgroup_insert(protin, txn->be->in);
+
+        if (txn->flags.ver < VER_2) {
+            /* stream client input directly to server */
+            serverout = txn->be->out;
+        }
+    }
+
+    n = prot_select(protin, PROT_NO_FD, &protout, NULL,
+                    timeout_sec ? &timeout : NULL);
+    if (n == -1 && errno != EINTR) {
+        syslog(LOG_ERR, "prot_select() failed in proxy_check_input(): %m");
+        fatal("prot_select() failed in proxy_check_input()", EX_TEMPFAIL);
+    }
+
+    if (n && protout) {
```

Comment:
Does this need to be `if (n > 0 && protout) {`?  If the prot_select above is interrupted then `n` could be -1 without fataling, which is truthy, and then the loop just below will iterate backwards from -1 to INT_MIN, underflow to INT_MAX, and keep looping until it reaches 0 (unless it crashes earlier than that due to an out of bounds access somewhere)

Reply:
Good catch.  This code is a specialized copy of proxy.c:proxy_check_input() which does the same thing.  However, it looks like protout will be NULL if prot_select( ) returns -1